snapshot/volume-snapshotting-proposal.md

```diff
@@ -0,0 +1,509 @@
+
+* For security purpose, if a snapshot object is created by a user, kubernetes should prevent other users duplicating this object in a different namespace if they happen to get the snapshot name.
+
+* There might be some existing snapshots taken by admins/users and they want to use those snapshots through kubernetes API interface.
+
+* **Create:**
+
+    1. The user creates a `VolumeSnapshot` referencing a persistent volume claim bound to a persistent volume
+
+    2. The controller fulfils the `VolumeSnapshot` by creating a snapshot using the volume plugins.
+
+    3. A new object `VolumeSnapshotData` is created to represent the actual snapshot binding the `VolumeSnapshot` with
+       the on-disk snapshot.
+
+* **List:**
+
+    1. The user is able to list all the `VolumeSnapshot` objects in the namespace
+
+* **Delete:**
+
+    1. The user deletes the `VolumeSnapshot`
+
+    2. The controller removes the on-disk snapshot. Note: snapshots have no notion of "reclaim policy" - there is
+       no way to recover the deleted snapshot.
+
+    3. The controller removes the `VolumeSnapshotData` object.
+
+* **Promote snapshot to PV:**
+
+    1. The user creates a persistent volume claim referencing the snapshot object in the annotation. The PVC must
+       belong to a `StorageClass` using the external volume snapshot provisioner.
+
+    2. The provisioner will use the `VolumeSnapshotData` object to create a persistent volume using the corresponding
+       volume snapshot plugin.
+
+    3. The PVC is bound to the newly created PV containing the data from the snapshot.
+
+
+The snapshot operation is a no-op for volume plugins that do not support snapshots via an API call (i.e. non-cloud storage).
+
+
+### API
+
+* The `VolumeSnapshot` object
+
+```
+// The volume snapshot object accessible to the user. Upon succesful creation of the actual
+// snapshot by the volume provider it is bound to the corresponding VolumeSnapshotData through
+// the VolumeSnapshotSpec
+type VolumeSnapshot struct {
+	metav1.TypeMeta `json:",inline"`
+	Metadata        metav1.ObjectMeta `json:"metadata"`
+
+	// Spec represents the desired state of the snapshot
+	// +optional
+	Spec VolumeSnapshotSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status represents the latest observer state of the snapshot
+	// +optional
+	Status VolumeSnapshotStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+type VolumeSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	Metadata        metav1.ListMeta  `json:"metadata"`
+	Items           []VolumeSnapshot `json:"items"`
+}
+
+// The desired state of the volume snapshot
+type VolumeSnapshotSpec struct {
+	// PersistentVolumeClaimName is the name of the PVC being snapshotted
+	// +optional
+	PersistentVolumeClaimName string `json:"persistentVolumeClaimName" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
```

Comment:
I feel using a VolumeSnapshotSource would be much easier to extend unless we are very sure PersistentVolumeClaim is the only source. 
For extension, maybe we can add source to represent some existing snapshots for the snapshots sharing across namespace use case.